ibis/backends/pyspark/compiler.py

```diff
@@ -1,2137 +1,502 @@
-    # We need to get the 'combined' time range that is a superset of every
-    # time context among child nodes, and pass this as context to
-    # source table to get all data within time context loaded.
-    arg_timecontexts = [
-        adjust_context(node, scope=scope, timecontext=timecontext)  # , **kwargs)
-        for node in op.selections
-        if timecontext
-    ]
-    adjusted_timecontext = combine_time_context(arg_timecontexts)
-    # If this is a sort or filter node, op.selections is empty
-    # in this case, we use the original timecontext
-    if not adjusted_timecontext:
-        adjusted_timecontext = timecontext
-    src_table = t.translate(
-        op.table, scope=scope, timecontext=adjusted_timecontext, **kwargs
-    )
-
-    col_in_selection_order = []
-    col_to_drop = []
-    result_table = src_table
-
-    for predicate in op.predicates:
-        col = t.translate(predicate, scope=scope, timecontext=timecontext, **kwargs)
-        # Due to an upstream Spark issue (SPARK-33057) we cannot
-        # directly use filter with a window operation. The workaround
-        # here is to assign a temporary column for the filter predicate,
-        # do the filtering, and then drop the temporary column.
-        filter_column = f"predicate_{guid()}"
-        result_table = result_table.withColumn(filter_column, col)
-        result_table = result_table.filter(F.col(filter_column))
-        result_table = result_table.drop(filter_column)
-
-    for selection in op.selections:
-        if isinstance(selection, ops.TableNode):
-            col_in_selection_order.extend(selection.schema.names)
-        elif isinstance(selection, ops.Value):
-            # If the selection is a straightforward projection of a table
-            # column from the root table itself (i.e. excluding mutations and
-            # renames), we can get the selection name directly.
-            if _can_be_replaced_by_column_name(selection, op.table):
-                col_in_selection_order.append(selection.name)
+@replace(p.Limit)
+def offset_to_filter(_):
+    # spark doesn't support dynamic limit, so raise an error if either limit or
+    # offset is not a literal expression
+    if isinstance(_.n, ops.Value) and _.n.find(ops.Relation):
+        raise com.UnsupportedOperationError(
+            "PySpark backend does not support dynamic limit."
+        )
+    if isinstance(_.offset, ops.Value) and _.offset.find(ops.Relation):
+        raise com.UnsupportedOperationError(
+            "PySpark backend does not support dynamic offset."
+        )
+    if _.offset == 0:
+        return _
+    # spark doesn't support offset by default, so we need to emulate it by first
+    # generating row numbers and then filtering out the first N rows
+    field_name = gen_name("ibis_row_number")
+    rel = _.parent.to_expr()
+    rel = rel.mutate(ibis.row_number().name(field_name))
+    rel = rel.filter(rel[field_name] > _.offset)
+    return _.copy(parent=rel, offset=0)
```

Comment:
This optimization is overall pretty interesting, and perhaps generalizable to other backends that don't support offset and/or dynamic limit. Good follow-up fodder (though not necessary by any means.)